### ISSUE TITLE
Design random query position generation for FRI

### ISSUE THREAD

**Issue by Al-Kindi-0** [0xPolygonMiden/miden-vm#753]
During the recursive STARK verification, the verifier needs to draw random integers which are the indices of the deep queries used in the FRI protocol. These integers belong to the range `0 ... lde_domain_size` and are drawn using the following method of a `RandomCoin<BaseField, RPO>`:
```rust
pub fn draw_integers(
        &mut self,
        num_values: usize,
        domain_size: usize,
    ) -> Result<Vec<usize>, RandomCoinError> {
        assert!(
            domain_size.is_power_of_two(),
            "domain size must be a power of two"
        );
        assert!(
            num_values < domain_size,
            "number of values must be smaller than domain size"
        );

        // determine how many bits are needed to represent valid values in the domain
        let v_mask = (domain_size - 1) as u64;

        // draw values from PRNG until we get as many unique values as specified by num_queries
        let mut values = Vec::new();
        for i in 0..1000 {
            // get the next pseudo-random value and read the first 8 bytes from it
            let bytes: [u8; 8] = self.next().as_bytes()[..8].try_into().unwrap();

            // convert to integer and limit the integer to the number of bits which can fit
            // into the specified domain
            let value = (u64::from_le_bytes(bytes) & v_mask) as usize;

            if values.contains(&value) {
                continue;
            }
            values.push(value);
            if values.len() == num_values {
                break;
            }
        }

        if values.len() < num_values {
            return Err(RandomCoinError::FailedToDrawIntegers(
                num_values,
                values.len(),
                1000,
            ));
        }

        Ok(values)
    }
```

This should be straight-forward to translate to Miden assembly code modulo the following remarks:

1. The loop's purpose is to keep drawing random indices until we get `num_values` **distinct** integers. How should we implement the `continue` and `break` for this loop?
2. Similar to the discussion in #751, we are using (at least) one permutation per random integer drawn. Since the `lde_domain_size` can not practically be larger than $2^{32} - 1$, does it make sense to modify the above method so that we draw $8$ random elements per permutation?

**Comment by bobbinth** [comment]
I think translating `break` and `continue` into if/else statements shouldn't be too difficult. But I am a bit worried bout checking uniqueness. Probably the simplest way to do it is to use a simple SMT of `log2(lde_domain_size)` to track already generated values - though I wonder if there is a better way to do it.



> Since the `lde_domain_size` can not practically be larger than $2^{32}−1$, does it make sense to modify the above method so that we draw 8 random elements per permutation?

I think so, but we'll need to think through how to make this change in Winterfell as well.

**Comment by Al-Kindi-0** [comment]
Another solution would be to use ND input to signal a duplicate `index`. In this case, the (offset) pointer for the position of the duplicate `index` is also provided and we just check that indeed the newly drawn `index` is equal to a previously drawn one.

**Comment by bobbinth** [comment]
I'm not sure nondeterminism would help us here. We can use it to prove to prove that something exists but not that something doesn't exist. For example, the prover would be able to claim that there are no duplicates when in fact there are.

Or am I missing something here?

**Comment by Al-Kindi-0** [comment]
In fact there are two problems:

1. Claiming that a drawn index is a a duplicate when it is not. This can be used in order to avoid a bad set of indices (from the prover's perspective).
2. Claiming that a duplicate is not a duplicate. This can be used in order to reduce the number of distinct queries.

It seems I was thinking of one of the problems but not the other.

Another solution might be to change the way we sample queries:

1. Sample distinct queries from the remainder codeword domain. Insuring distinctness when the length of the domain is 64 should be straightforward. 
2. For each `q` of the queries in point 1:
        1.  Draw a random integer between 1 and the number of FRI layers $n$. Call it $i$.
        2. Compute a query on the lde-domain as $q\tau^i$ where $\tau$ is an $n$-th root of unity.

**Comment by Al-Kindi-0** [comment]
Partially closed by #937 with the issue of handling duplicate queries moved to #941
